frontends/smvparser.y

```diff
@@ -511,33 +665,62 @@ basic_expr: constant {
                   smt::Term res = enc.solver_->make_term(smt::Mod, a->getTerm(), b->getTerm());
                   $$ = new SMVnode(res);
               }else{
-                  smt::Term res = enc.solver_->make_term(smt::BVSmod, a->getTerm(), b->getTerm());
-                  $$ = new SMVnode(res);
+                  SMVnode::BVtype bvs_a = a->getBVType();
+                  SMVnode::BVtype bvs_b = b->getBVType();
+                  if (bvs_a == bvs_b == SMVnode::Unsigned){
+                    smt::Term res = enc.solver_->make_term(smt::BVUrem, a->getTerm(), b->getTerm());
+                    $$ = new SMVnode(res,SMVnode::Unsigned);
+                  } else if (bvs_a == bvs_b == SMVnode::Signed){
+                    smt::Term res = enc.solver_->make_term(smt::BVSmod, a->getTerm(), b->getTerm());
+                    $$ = new SMVnode(res,SMVnode::Signed);
+                  } else{
+                    throw CosaException ("Unsigned/Signed unmatch");
+                  } 
               }
             }
             | basic_expr OP_SHIFTR basic_expr{
               SMVnode *a = $1;
               SMVnode *b = $3;
               smt::Term res = enc.solver_->make_term(smt::BVLshr, a->getTerm(), b->getTerm());
-              $$ = new SMVnode(res);
+              $$ = new SMVnode(res,a->getBVType());
             }
             | basic_expr OP_SHIFTL basic_expr{
               SMVnode *a = $1;
               SMVnode *b = $3;
               smt::Term res = enc.solver_->make_term(smt::BVShl, a->getTerm(), b->getTerm());
-              $$ = new SMVnode(res);
+              $$ = new SMVnode(res,a->getBVType());
             }
             | basic_expr sizev {
               throw CosaException("No word1");
             }
             | basic_expr OP_CON basic_expr  {
               SMVnode *a = $1;
               SMVnode *b = $3;
-              smt::Term res = enc.solver_->make_term(smt::Concat, a->getTerm(), b->getTerm());
-              $$ = new SMVnode(res);
+              SMVnode::BVtype bvs_a = a->getBVType();
+              SMVnode::BVtype bvs_b = b->getBVType();
+              if(bvs_a != bvs_b){
+                throw CosaException("Unsigned/Signed unmatch");
+              } else{
+                smt::Term res = enc.solver_->make_term(smt::Concat, a->getTerm(), b->getTerm());
+                $$ = new SMVnode(res,SMVnode::Unsigned);
+              }               
+            }
+            | basic_expr "[" integer_val "]" {
+                throw CosaException("No index Subscript");
             }
             | basic_expr "[" basic_expr ":" basic_expr "]"{
```

Comment:
It might be better to be more specific here (if that's not too difficult). The extract operator has to take non-negative integers (just saying integers would be fine). So instead of `basic_expr`, I think it could be `integer_constant`?

Reply:
I think in the manual, it says the type of basic_expr should be an array? Do you want me to change `basic_expr` to `array`?

Reply:
No I wouldn't -- there's a bit of a terminology mismatch here. What they're calling an array, we would call a bit-vector (an array of bits). An array in our terminology is an SMT array -- it's a data structure that can be defined over arbitrary index and element sorts (e.g. an array from integers to bit-vectors). So what I would actually expect to see is:
```
| bitvec_expr "[" integer_constant ":" integer_constant "]"
```

but you don't need to actually have a dedicated bit vector expression. I think it's probably fine as-is, but I'm curious how it fails if you did something like this:

```
VAR b : unsigned word[8];
VAR t : boolean;
INIT b[t:t] = 0ud1_1
```

because `b[t:t]` is not supported, it would have to be integer values like `b[0:0]`

Reply:
actually I see that you check they're integers below.